### PR TITLE
Add code action to Qualify as Path

### PR DESCRIPTION
**Pull request by sdankel**
## Description

Closes https://github.com/FuelLabs/sway/issues/5189

## Checklist

- [x] I have linked to any relevant issues.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [ ] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [ ] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by JoshuaBatty** [comment]
Nice, excited to see this land. I just pulled down to test locally and for some reason, I'm not seeing any code lens icons when I'm over an item that needs importing/qualifying.... Not sure why this is the case. Just double checking it's working for you?

**Comment by sdankel** [comment]
> Nice, excited to see this land. I just pulled down to test locally and for some reason, I'm not seeing any code lens icons when I'm over an item that needs importing/qualifying.... Not sure why this is the case. Just double checking it's working for you?

Yes, it's definitely working for me. Are you not seeing the hover modal at all, or just no quick fixes?

image
image

**Review by JoshuaBatty — APPROVED** [review]
Managed to test this locally. Works great. 🚀

**Review by xunilrj — APPROVED** [review]

File: sway-lsp/src/capabilities/code_actions/diagnostic/mod.rs [hunk 1]
```diff
@@ -1,24 +1,27 @@
...
-use super::CODE_ACTION_IMPORT_TITLE;
+use self::qualify::qualify_code_action;
 
 /// Returns a list of [CodeActionOrCommand] based on the relavent compiler diagnostics.
 pub(crate) fn code_actions(ctx: &CodeActionContext) -> Option<Vec<CodeActionOrCommand>> {
     // Find diagnostics that have attached metadata.
     let diagnostics_with_data = ctx.diagnostics.iter().filter_map(|diag| {
         if let Some(data) = diag.clone().data {
-            return serde_json::from_value::<DiagnosticData>(data).ok();
+            if let Ok(data) = serde_json::from_value::<DiagnosticData>(data) {
```
[the remark this answers is not in the document]
Comment (cr-fuel) [reply sway-lsp/src/capabilities/code_actions/diagnostic/mod.rs:15]: The following line is shorter:

```
return serde_json::from_value::<DiagnosticData>(data).map(|d| (diag.rnge, data)).ok();
```

**Review by cr-fuel — APPROVED** [review]

File: sway-lsp/src/capabilities/code_actions/diagnostic/mod.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdankel) [reply sway-lsp/src/capabilities/code_actions/diagnostic/mod.rs:15]: Good catch, I will update this in a future PR so as not to wait for approvals again 😄

File: sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+                ..Default::default()
+            })
+        })
+        .collect::<Vec<_>>();
+
+    if !actions.is_empty() {
+        return Some(actions);
+    }
+
+    None
```
Comment (kayagokalp) [inline sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs:54]:
```suggestion
actions.into();
```

You can call `.into()` for `Vec` to convert it into a `Option<Vec<_>>` :) rust std has really nice from implementations in place

File: sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+                kind: Some(CodeActionKind::QUICKFIX),
+                edit: Some(WorkspaceEdit {
+                    changes: Some(changes),
+                    ..Default::default()
+                }),
+                data: Some(Value::String(ctx.uri.to_string())),
+                ..Default::default()
+            })
+        })
+        .collect::<Vec<_>>();
```
Comment (kayagokalp) [inline sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs:48]:
```suggestion
        .collect::<Option<Vec<_>>>();
```

Might work as we have: https://doc.rust-lang.org/std/option/enum.Option.html#method.from_iter

File: sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs (hunk 2 above)
Comment (kayagokalp) [reply sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs:54]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f22c9270f9571ffde4439c6e0db51740

**Review by kayagokalp — COMMENTED** [review]
Left a small nit, looks good to me though 👍

File: sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs (hunk 2 above)
Comment (sdankel) [reply sway-lsp/src/capabilities/code_actions/diagnostic/qualify.rs:54]: Thanks for the playground link. My intent here is to return None when the vector is empty. It looks like calling `into` on an empty vec returns `Some([])`, which is not quite what I want. However, I'm not sure if it makes a difference from the user's perspective.

**Pull request merged**